R/getDominantTaxa.R

```diff
@@ -0,0 +1,152 @@
+#' Get Dominant Taxa
+#'
+#' This function return the most dominant taxa.
+#'
+#' @param x
+#' A \code{\link[SummarizedExperiment:SummarizedExperiment-class]{SummarizedExperiment}}
+#' object
+#'
+#' @param rank A single character defining a taxonomic rank. Must be a value of
+#'   \code{taxonomicRanks()} function.
+#'
+#' @param group With group, it is possible group observations in overview.
+#'
+#' @param name A name for the column of the colData where the dominant taxa
+#' will be stored in.
+#'
+#' @details
+#' \code{getDominantTaxa} extracts the most abundant taxa
+#' in a \code{\link[=SummarizedExperiment-class]{SummarizedExperiment}} object.
+#'
+#' @return \code{x} with additional \code{\link{colData}} named
+#'   \code{*name*}
+#'
+#' @name getDominantTaxa
+#' @export
+#'
+#' @author Leo Lahti and Tuomas Borman. Contact: \url{microbiome.github.io}
+#'
+#' @examples
+#' data(GlobalPatterns)
+#' x <- GlobalPatterns
+#'
+#' #Finds the dominant taxa.
+#' x <- getDominantTaxa(x)
+#' #Information is stored to colData
+#' colData(x)
+#'
+#' #If taxonomic information is available, it is possible to find the most dominant
+#' #group from specific taxonomic level, here family level. The name of column can be specified.
+#' x <- getDominantTaxa(x, rank="Family", name="Dominant Family")
+#' colData(x)
+#'
+#' x <- microbiomeDataSets::dietswap()
+#' #With group, it is possible to group observations based on group specified by user
+#' x <- getDominantTaxa(x, group = "nationality")
+#'
+NULL
+
+#' @rdname getDominantTaxa
+#' @export
+setGeneric("getDominantTaxa",signature = c("x"),
+           function(x, rank = NULL, group = NULL, name = "dominant_taxa")
+               standardGeneric("getDominantTaxa"))
+
+
+#' @rdname getDominantTaxa
+#' @export
+setMethod("getDominantTaxa", signature = c(x = "SummarizedExperiment"),
+          function(x, rank = NULL, group = NULL, name = "dominant_taxa"){
+
+              #Input check
+              #rank check
+              if(!is.null(rank)){
+                  if(!.is_a_string(rank)){
+                      stop("'rank' must be an single character value.",
+                           call. = FALSE)
+                  }
+                  .check_taxonomic_rank(rank, x)
+              }
+
+              #group check
+              if(!is.null(group)){
+                  if(isFALSE(any(group %in% colnames(colData(x))))){
+                      stop("'group' variable must be in colnames(colData(x))")
+                  }
+              }
+
+              #name check
+              if(!.is_non_empty_string(name)){
+                  stop("'name' must be a non-empty single character value.",
+                       call. = FALSE)
+              }
+
+              #If "rank" is not NULL, species are aggregated according to the taxonomic rank
+              #that is specified by user.
+              if (!is.null(rank)) {
+                  x <- agglomerateByRank(x, rank = rank)
+              }
```

Comment:
you change the input, but return it. I think this not the function of this function. The user can choose to do it themself by using `getDominantTaxa(agglomerateByRank(x, rank = rank), rank = rank)`

An alternative would be:

```
tmp <- agglomerateByRank(x, rank = rank)
```

Better would be to restructure `agglomerateByRank` and add a method for the signature `ANY` assuming `x` is in this case a `matrix`-like object. Then you could do something like this

```
mat <- agglomerateByRank(assay(x, abund_values), rank_data = rank_data)
```

However, I think this is to complicated for now. Instead reuse some code and try this 

```
col <- which( taxonomyRanks(x) %in% rank )
tax_factors <- .get_tax_groups(x, col = col, onRankOnly = FALSE)
mat <- mergeRows(assay(x, abund_values), f = tax_factors)
```

Reply:
Changed to 

# Selects the level
                  col <- which( taxonomyRanks(x) %in% rank )

                  # Function from taxonomy.R. Divides taxas to groups where they belong.
                  tax_factors <- .get_tax_groups(x, col = col, onRankOnly = FALSE)

                  # Merges abundances within the groups
                  mat <- mergeRows(x, f = tax_factors)